### PR TITLE
Add SSH command to CLI

### PR DESCRIPTION
**Pull request by nicolai86**
## What does this PR do?

This PR implements `lightning ssh`, a convenient wrapper around running the correct SSH invocation for flow/ works.

Specifically:

```bash
❯ lightning ssh
[?] What app to SSH into?: example
 > example
 > example2

[?] Which component to SSH into?: flow
   root.my_work
 > flow
Welcome to Ubuntu 20.04.5 LTS (GNU/Linux 5.4.209-116.367.amzn2.x86_64 x86_64)

 * Documentation:  https://help.ubuntu.com
 * Management:     https://landscape.canonical.com
 * Support:        https://ubuntu.com/advantage

This system has been minimized by removing packages and content that are
not required on a system that users do not log into.

To restore this content, you can run the 'unminimize' command.

The programs included with the Ubuntu system are free software;
the exact distribution terms for each program are described in the
individual files in /usr/share/doc/*/copyright.

Ubuntu comes with ABSOLUTELY NO WARRANTY, to the extent permitted by
applicable law.

zeus@lightningapp-01gg7wqgq8fwxptmeyn4s9mq7z:~$
```

You can specify the app to short-circuit selection:

```
❯ lightning ssh --app-name test-app
[?] Which component to SSH into?: flow
   root.my_work
 > flow
```

You can specify the component:

```
❯ lightning ssh --app-name test-app --component-name root.my_work
```

Because `click` doesn't offer an interactive dialog I've added `inquirer` as dependency, which allows me to build an ok screen prompt for user interactions.

Fixes #\<issue_number>

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda

### CODE REVIEW

File: src/lightning_app/cli/lightning_cli.py [hunk 1]
```diff
@@ -337,6 +339,75 @@ def stop() -> None:
...
 
 
+@_main.command("ssh")
+@click.option(
+    "--app-id",
+    "app_id",
+    type=str,
+    default=None,
+    required=False,
+)
```
Comment (awaelchli) [inline src/lightning_app/cli/lightning_cli.py:349]: In the PR description you have the example where inquirer asks for the app name to connect to. 
I would suggest we also make it available here. Using the name seems the most intuitive to me, especially because the name is displayed in many places (CLI when you start app, in the admin front and center, etc.).

File: src/lightning_app/cli/lightning_cli.py [hunk 2]
```diff
@@ -337,6 +339,75 @@ def stop() -> None:
...
+@click.option(
+    "--component-name",
+    "component_name",
+    type=str,
+    default=None,
+    help="Specify which component to SSH into",
+)
+def ssh(app_id: str = None, component_name: str = None) -> None:
+    """SSH into a Lighting App."""
+    import inquirer
```
Comment (awaelchli) [inline src/lightning_app/cli/lightning_cli.py:359]: Let's move it to the top?

File: src/lightning_app/cli/lightning_cli.py [hunk 3]
```diff
@@ -337,6 +339,75 @@ def stop() -> None:
...
+)
+@click.option(
+    "--component-name",
+    "component_name",
+    type=str,
+    default=None,
+    help="Specify which component to SSH into",
+)
+def ssh(app_id: str = None, component_name: str = None) -> None:
+    """SSH into a Lighting App."""
```
Comment (awaelchli) [inline src/lightning_app/cli/lightning_cli.py:358]:
```suggestion
    """SSH into a Lightning App."""
```

File: src/lightning_app/cli/lightning_cli.py [hunk 4]
```diff
@@ -337,6 +339,75 @@ def stop() -> None:
...
+    component_id = None
+    if component_name == "flow":
+        component_id = f"lightningapp-{app_id}"
+    elif component_name is not None:
+        work_id = next(iter([work.id for work in components if work.name == component_name]), None)
+        if work_id is not None:
+            component_id = f"lightningwork-{work_id}"
+
+    if component_id is None:
+        raise click.ClickException(f"unable to find app component with name {component_name}")
```
Comment (awaelchli) [inline src/lightning_app/cli/lightning_cli.py:402]: Is it possible to list the components the app has as part of this error message? To help the user fix a typo or generally show what choices they have.

File: src/lightning_app/cli/lightning_cli.py [hunk 5]
```diff
@@ -337,6 +339,75 @@ def stop() -> None:
...
+                choices=[work.name for work in components] + ["flow"],
+            )
+        ]
+        component_name = inquirer.prompt(available_components)["component_name"]
+
+    component_id = None
+    if component_name == "flow":
+        component_id = f"lightningapp-{app_id}"
+    elif component_name is not None:
+        work_id = next(iter([work.id for work in components if work.name == component_name]), None)
```
Comment (awaelchli) [inline src/lightning_app/cli/lightning_cli.py:397]:
```suggestion
        work_id = next((work.id for work in components if work.name == component_name), None)
```
Using Python generators, the list and iter are not necessary.

File: src/lightning_app/cli/lightning_cli.py [hunk 6]
```diff
@@ -337,6 +339,75 @@ def stop() -> None:
...
+)
+def ssh(app_id: str = None, component_name: str = None) -> None:
+    """SSH into a Lighting App."""
+    import inquirer
+
+    app_manager = _AppManager()
+    if app_id is None:
+        apps = app_manager.list_apps(phase_in=[V1LightningappInstanceState.RUNNING])
+        if len(apps) == 0:
+            raise click.ClickException("no running apps available.")
```
Comment (awaelchli) [inline src/lightning_app/cli/lightning_cli.py:365]: What could we recommend to the user for the next step? Point them to the webpage with the list of the apps they could start? We could let them interactively choose an app to start using the inquirer, but one might have a very long list of stopped apps. We could list the top K recently used apps.

File: src/lightning_app/cli/lightning_cli.py [hunk 7]
```diff
@@ -337,6 +339,75 @@ def stop() -> None:
...
+                choices=[app.name for app in apps],
+            ),
+        ]
+        app_name = inquirer.prompt(available_apps)["app_name"]
+        app_id = [app.id for app in apps if app.name == app_name][0]
+
+    try:
+        instance = app_manager.get_app(app_id=app_id)
+    except ApiException:
+        raise click.ClickException("failed fetching app instance")
```
Comment (awaelchli) [inline src/lightning_app/cli/lightning_cli.py:380]: What is the action the user should take? Do they need to log in first?

File: src/lightning_app/cli/lightning_cli.py [hunk 8]
```diff
@@ -337,6 +339,75 @@ def stop() -> None:
...
+        if work_id is not None:
+            component_id = f"lightningwork-{work_id}"
+
+    if component_id is None:
+        raise click.ClickException(f"unable to find app component with name {component_name}")
+
+    app_cluster = app_manager.get_cluster(cluster_id=instance.spec.cluster_id)
+    ssh_endpoint = app_cluster.status.ssh_gateway_endpoint
+
+    os.execv("/usr/bin/ssh", ["-tt", f"{component_id}@{ssh_endpoint}"])
```
Comment (awaelchli) [inline src/lightning_app/cli/lightning_cli.py:407]: Could we do some research how to find the `ssh` command? IMO, we should at least check if the ssh command is available. For example with: `ssh_available = subprocess.call(["command", "-v", "ssh"]) == 0`.

File: src/lightning_app/cli/lightning_cli.py [hunk 9]
```diff
@@ -337,6 +339,75 @@ def stop() -> None:
...
+    pass
```
Comment (awaelchli) [inline src/lightning_app/cli/lightning_cli.py:408]:
```suggestion
```

**Review by awaelchli — CHANGES_REQUESTED** [review]

**Comment by nicolai86** [comment]
@Borda yes docs are [here](https://github.com/Lightning-AI/lightning/pull/15316)

File: src/lightning_app/cli/lightning_cli.py (hunk 2 above)
Comment (nicolai86) [reply src/lightning_app/cli/lightning_cli.py:359]: moved to the top :+1:

File: src/lightning_app/cli/lightning_cli.py (hunk 4 above)
Comment (nicolai86) [reply src/lightning_app/cli/lightning_cli.py:402]: good point, I've added that.

File: src/lightning_app/cli/lightning_cli.py (hunk 7 above)
Comment (nicolai86) [reply src/lightning_app/cli/lightning_cli.py:380]: the user selected an app that doesn't exist- when I change the interface to accept an app_name this error message will be more descriptive.

File: src/lightning_app/cli/lightning_cli.py (hunk 6 above)
Comment (nicolai86) [reply src/lightning_app/cli/lightning_cli.py:365]: For now I'll reword this to include a suggestion to start a Lightning App. Let's add more functionality once we see people using this feature.

File: src/lightning_app/cli/lightning_cli.py (hunk 8 above)
Comment (nicolai86) [reply src/lightning_app/cli/lightning_cli.py:410]: good point, I meant to switch this over to [shutil](https://docs.python.org/3/library/shutil.html)

File: src/lightning_app/cli/lightning_cli.py (hunk 1 above)
Comment (nicolai86) [reply src/lightning_app/cli/lightning_cli.py]: fair point. I've reworked this and the docs to accept `app-name` instead.

File: src/lightning_app/cli/lightning_cli.py [hunk 10]
```diff
@@ -341,6 +345,86 @@ def stop() -> None:
...
+                message="What app to SSH into?",
+                choices=available_app_names,
+            ),
+        ]
+        app_name = inquirer.prompt(available_apps)["app_name"]
+    app_id = next((app.id for app in apps if app.name == app_name), None)
+    if app_id is None:
+        raise click.ClickException(
+            f"unable to find running app with name {app_name}. " + f"Available running apps are {available_app_names}"
+        )
```
Comment (awaelchli) [inline src/lightning_app/cli/lightning_cli.py:385]: Nit: formatting
```suggestion
        raise click.ClickException(
            f"Unable to find a running app with name {app_name} in your account. Available running apps are: {', '.join(available_app_names)}"
        )
```

File: src/lightning_app/cli/lightning_cli.py [hunk 11]
```diff
@@ -341,6 +345,86 @@ def stop() -> None:
...
+    default=None,
+    help="Specify which component to SSH into",
+)
+def ssh(app_name: str = None, component_name: str = None) -> None:
+    """SSH into a Lightning App."""
+
+    app_manager = _AppManager()
+    apps = app_manager.list_apps(phase_in=[V1LightningappInstanceState.RUNNING])
+    if len(apps) == 0:
+        raise click.ClickException("no running apps available. Start a Lightning App in the cloud to use this feature.")
```
Comment (awaelchli) [inline src/lightning_app/cli/lightning_cli.py:369]: Super nit:
```suggestion
        raise click.ClickException("No running apps available. Start a Lightning App in the cloud to use this feature.")
```

File: src/lightning_app/cli/lightning_cli.py [hunk 12]
```diff
@@ -341,6 +345,86 @@ def stop() -> None:
...
+    elif component_name is not None:
+        work_id = next((work.id for work in components if work.name == component_name), None)
+        if work_id is not None:
+            component_id = f"lightningwork-{work_id}"
+
+    if component_id is None:
+        raise click.ClickException(
+            f"unable to find app component with name {component_name}. "
+            + f"Available components are {available_component_names}"
+        )
```
Comment (awaelchli) [inline src/lightning_app/cli/lightning_cli.py:415]: Super nit:
```suggestion
        raise click.ClickException(
            f"Unable to find an app component with name {component_name}. "
            f"Available components are: {', '.join(available_component_names)}"
        )
```

File: src/lightning_app/cli/lightning_cli.py [hunk 13]
```diff
@@ -341,6 +345,86 @@ def stop() -> None:
...
+
+    app_cluster = app_manager.get_cluster(cluster_id=instance.spec.cluster_id)
+    ssh_endpoint = app_cluster.status.ssh_gateway_endpoint
+
+    ssh_path = shutil.which("ssh")
+    if ssh_path is None:
+        raise click.ClickException(
+            "unable to find ssh binary. You must install the ssh binary to use this functionality"
+        )
```
Comment (awaelchli) [inline src/lightning_app/cli/lightning_cli.py:424]: Nit:
```suggestion
        raise click.ClickException(
            "Unable to find the ssh binary. You must install ssh first to use this functionality."
        )
```

**Review by awaelchli — APPROVED** [review]
Thanks for addressing my review comments earlier!

**Comment by nicolai86** [comment]
> Thanks for addressing my review comments earlier!

thanks for the review!!

**Review by manskx — APPROVED** [review]
